rlmolecule/alphazero/tfalphazero_problem.py

```diff
@@ -0,0 +1,272 @@
+
+        # Get the flat value and prior_logit predictions
+        flat_values_logits, flat_prior_logits = self.policy_model(flattened_inputs)
+
+        # We put the parent node first in our batch inputs, so this slices
+        # the value prediction for the parent
+        value_preds = tf.reshape(flat_values_logits, [batch_size, max_actions_per_node, -1])[:, 0, 0]
+
+        # Next we get a mask to see where we have valid actions and replace priors for
+        # invalid actions with negative infinity (these get zeroed out after softmax).
+        # We also only return prior_logits for the child nodes (not the first entry)
+        mask_values = [self._input_masks[inp.name] for inp in self._policy_model.inputs]
+        action_mask = make_action_mask(inputs, mask_values)
+        prior_logits = tf.reshape(flat_prior_logits, [batch_size, max_actions_per_node])
+        masked_prior_logits = tf.where(
+            action_mask, 
+            prior_logits,
+            tf.ones_like(prior_logits) * prior_logits.dtype.min)[:, 1:]
+
+        return value_preds, masked_prior_logits
+
+    @property
+    def input_masks(self):
+        return self._input_masks
+
+
+class TFAlphaZeroProblem(AlphaZeroProblem):
+    def __init__(self,
+                 engine: sqlalchemy.engine.Engine,
+                 model: tf.keras.Model,
+                 model_input_masks: dict = {},
+                 policy_checkpoint_dir: Optional[str] = None,
+                 **kwargs
+                 ) -> None:
+
+        super(TFAlphaZeroProblem, self).__init__(engine, **kwargs)
+        #self.policy_model = build_policy_trainer(hallway_size, hidden_layers, hidden_dim)
+        self.model_input_masks = model_input_masks
+        self.policy_model = build_policy_trainer(model)
+        self.policy_checkpoint_dir = policy_checkpoint_dir
+        policy_model_layer = self.policy_model.layers[-1].policy_model
+        self.policy_evaluator = tf.function(experimental_relax_shapes=True)(policy_model_layer.predict_step)
+        self._checkpoint = None
+
+    def initialize_run(self):
+        """
+        Load the most recent policy checkpoint
+        """
+        super().initialize_run()
+
+        if self.policy_checkpoint_dir:
+            new_checkpoint = tf.train.latest_checkpoint(self.policy_checkpoint_dir)
+            if new_checkpoint != self._checkpoint:
+                self._checkpoint = new_checkpoint
+                status = self.policy_model.load_weights(self._checkpoint)
+                status.assert_existing_objects_matched()
+                logger.info(f'Loaded checkpoint {self._checkpoint}')
+            elif new_checkpoint == self._checkpoint:
+                logger.info(f'Skipping already loaded {self._checkpoint}')
+            else:
+                logger.info('No checkpoint found')
+
+    @abstractmethod
+    def _get_network_inputs(self, state: GraphSearchState) -> Dict:
```

Comment:
we probably don't want the leading underscore if this is a public method implemented by the user. Is 'get_network_inputs` intuitive? Or maybe `get_policy_inputs`?

Reply:
yeah, agreed.  we use policy most places (rather than "network") so that would be better.